### ISSUE TITLE
No instant streaming

### ISSUE THREAD

**Issue by RichardHoman** [maennchen/ZipStream-PHP#94]
# Description of the problem
I may have been misinformed or I have misread. But i believe that this lib is meant to stream the zip to the client/browser on the fly. However. When I try to stream a zip consisting of a few large files, streaming does not start immediately. The browser just waits for data.

# Example code
I've used the example code as placed here on github.

# Informations
The script does function. eventually a proper zip will start to download. the only problem is that it does not start streaming immediately, which means that the file is being saved temporarily on the server somehow - which is not going to work since in the production environment it will need to serve zip files of over 32GB in size.

* ZipStream-PHP version: 1.0.0
* PHP version: 7.2

**Comment by NicolasCARPi** [comment]
Hello, can you confirm this on multiple browsers?

**Comment by RichardHoman** [comment]
> Hello, can you confirm this on multiple browsers?

Confirmed on Chrome, Firefox and Edge. (All most recent versions.)

One more thing: Since I don't use composer I used the 1.0.0 version from [PHP-download.com](https://php-download.com/package/maennchen/zipstream-php)

**Comment by NicolasCARPi** [comment]
There has been non trivial changes since the tagged 1.0.0-alpha, and I think php-download.com gives you the tagged version. Can you please pull from master and try again?

**Comment by RichardHoman** [comment]
Thank you for your reply. I've pulled the latest master. Sadly this did not solve the issue.

For testing purposes I'm trying to zip/stream 2 existing zip files, both around 6GB in size. The request from Chrome is pending and fails with a 500 error after a while. The error log shows an obvious error:
`PHP Fatal error:  Maximum execution time of 30 seconds exceeded in /mnt/                            opslag/bestanden.zonlicht.eu/zst/vendor/maennchen/zipstream-php/src/Stream.php o                            n line 223`

Apart from this, there are no errors. This specific shoudn't prevent the stream for initially starting and running for the first 30 seconds at least.

**Comment by maennchen** [comment]
@RichardHoman What is configured for `output_buffering` in your PHP settings?

The maximum execution time of 30 seconds can easily be reached by 12 Gb, please set the `max_execution_time` accordingly.

**Comment by RichardHoman** [comment]
Good day Männchen,

Output buffering is set to 4096 in php.ini
I've changed to max_execution_time to make sure this will not be a problem in the future when the current issue is fixed.

Best regards,
Richard

**Comment by maennchen** [comment]
@RichardHoman Please set `output_buffering` to 0 and try again. If the library really is not working as intended, you'd exceed PHP's memory. If you get a file with no errors, it's ok.

**Comment by RichardHoman** [comment]
Thanks. All these changes later I'm actually not getting a zip anymore.. now the browser just outputs `PKz�vM�,��!!	hello.txt��,V���T�����Լ�b��4��Ԝ�|���PKz�vM�,��!!	hello.txtPK7H`.

>> See: https://bestanden.zonlicht.eu/zst/www/index.php

This is the current code in the test script:
`<?php
	$zipfilename = "voorbeeld.zip";
	require '../vendor/autoload.php';
	$zip = new ZipStream\ZipStream($zipfilename);
	$zip->addFile('hello.txt', 'This is the contents of hello.txt');
	$zip->finish();
`
I'm pretty much lost by now..

**Comment by NicolasCARPi** [comment]
> Thanks. All these changes later I'm actually not getting a zip anymore.. now the browser just outputs `PK����z�vM�,��!! hello.txt���,V���T�����Լ�b��4��Ԝ�|����PK������z�vM�,��!! hello.txtPK����7H`.

That's the zipfile being displayed as ASCII in the browser. It means the browser didn't recognize the stream as a zip file. You can fix this like that:

~~~php
<?php
use ZipStream\Option\Archive as ArchiveOptions;

require 'vendor/autoload.php';
$zipfilename = "voorbeeld.zip";
$opt = new ArchiveOptions();
$opt->setSendHttpHeaders(true);
$opt->setHttpHeaderCallback('header');
$zip = new ZipStream\ZipStream($zipfilename, $opt);
$zip->addFile('hello.txt', 'This is the contents of hello.txt');
$zip->finish();
~~~

**Comment by RichardHoman** [comment]
Thank you for this helpful explanation. The script works now. There is just one last question I believe you can answer.

After opening the testpage it takes 75 seconds for the zip to start downloading. I believe this is due to the fact that the first actual file added to the zip is fairly big (+/- 7GB). My first question is whether this is the right assumption. The second question is if there is a way to start pushing the zip to the browser, even during the processing of that file.

In other words: It seems the script is loading the file, compressing the whole thing and only then starts to send it to the browser. Can the file be compressed and sent in 'chunks' so to say?

Best regards,
Richard

**Comment by NicolasCARPi** [comment]
The default method is STORE, not DEFLATE, so the file is actually not compressed. I believe the initial time might be needed by the server to just read the file entirely so it can do **magic stuff**, and then start sending it. So this delay might not be skippable. But if you add a nice message "Please wait, this may take some time…", it'll be okay ;)

**Comment by RichardHoman** [comment]
Thanks for your reply. I thought this was the case. 

All in all ZipStream-PHP works like a charm, so thanks again!
